Implement WebGL timer queries extension (EXT_disjoint_timer_query)

gouletr opened the pull request:
As discussed in #4572, to use timer queries (as opposed to other queries) we are required to use the extension functions, otherwise using timer related queries on standard queries functions result in WebGL error. Also, the extension [EXT_disjoint_timer_query](https://www.khronos.org/registry/webgl/extensions/EXT_disjoint_timer_query/) is available for both WebGL1&2, so it was implemented as such.

gouletr commented:
`getQueryObjectEXT` returns a `GLuint64EXT` for `QUERY_RESULT_EXT`, I'm not sure I got that right?

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1411,97 @@ var LibraryGL = {
     GLctx.bufferSubData(target, offset, HEAPU8.subarray(data, data+size));
   },
 
+  // Queries EXT
```
This addition duplicates all of the query-related code. Let's remove this and move the query functions outside `#if USE_WEBGL2=1` guard.

Then to get these working via the `EXT` suffixes, use routing functions like this:

    glGenQueriesEXT: 'glGenQueries',
    glDeleteQueriesEXT: 'glDeleteQueries',
    glIsQueryEXT: 'glIsQuery',
    glBeginQueryEXT: 'glBeginQuery',
    glEndQueryEXT: 'glEndQuery',
    glGetQueryivEXT: 'glGetQueryiv',
    glGetQueryObjectuivEXT: 'glGetQueryObjectuiv',

The `glGenQueriesEXT__sig` should not then be needed, since the compiler should know that `glGenQueriesEXT__sig == glGenQueries__sig` when it is doing the routing.

gouletr replied:
Yeah I duplicated it because I am not sure if we can overwrite the regular one with the extension one for WebGL2 and expect it to work for both regular and timer queries.

juj replied:
That's a good distinction to make, we don't want to assume that.

juj replied:
Thinking about how to structure this, it will be silly if one needs to create two different types of query objects, "EXT query objects" and "WebGL 2 query objects" if one wants to do both types of queries in an application.

Let's pause here for a second, I'll raise this up to the WebGL spec mailing list and we'll see if there can be some consensus.

gouletr replied:
Indeed, but they look different from the spec: `WebGLQuery` versus `WebGLTimerQueryEXT`.

juj replied:
Alright, looks like duplicating all the functions is, sadly, the best way to go here, we do want to strongly segregate the APIs and objects here. In the future if the WebGL spec extensions end up explicitly noting that this doesn't need to be the case, then we can fuse them together. Great work for being so specific with this aspect, this would have fallen through my attention completely otherwise.

gouletr commented on src/library_gl.js:
```diff
@@ -614,6 +614,17 @@ var LibraryGL = {
 
+      context.disjointTimerQueryExt = GLctx.getExtension("EXT_disjoint_timer_query");
+      if (context.disjointTimerQueryExt) {
+        GLctx['createQueryEXT'] = function() { return context.disjointTimerQueryExt['createQueryEXT'](); };
```
I haven't tested if we can use non-timer related values with EXT versions of the API, not sure if that works? e.g. `ext.beginQuery(gl.ANY_SAMPLES_PASSED);` So you think we can replace regular versions with the EXT one and it should still work?

juj replied:
Oh, sorry, I am mistaken here. I thought that WebGL 2 adopted EXT_disjoint_timer_query into the core WebGL 2 specification, but looking closer now, I realize that it did not, and the supported query functionality is separate. Let me re-read this whole thing from this updated perspective.

juj replied:
Let's keep only the single line `context.disjointTimerQueryExt = GLctx.getExtension("EXT_disjoint_timer_query");` from this block. That is, obtain the extension but not do the routing functions, since they are a bit redundant.

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1411,97 @@ var LibraryGL = {
+  glGenQueriesEXT__sig: 'vii',
+  glGenQueriesEXT: function(n, ids) {
+    for (var i = 0; i < n; i++) {
+      var query = GLctx['createQueryEXT']();
```
Let's call `context.disjointTimerQueryExt.createQueryEXT()` here directly now that we don't do the routing stubs.

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1411,97 @@ var LibraryGL = {
+      var id = {{{ makeGetValue('ids', 'i*4', 'i32') }}};
+      var query = GL.queries[id];
+      if (!query) continue; // GL spec: "unused names in ids are ignored, as is the name zero."
+      GLctx['deleteQueryEXT'](query);
```
`context.disjointTimerQueryExt.deleteQueryEXT()`

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1411,97 @@ var LibraryGL = {
+      }
+      var id = GL.getNewId(GL.queries);
+      query.name = id;
+      GL.queries[id] = query;
```
`GL.timerQueriesEXT[id] = query;`

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1411,97 @@ var LibraryGL = {
+  glDeleteQueriesEXT: function(n, ids) {
+    for (var i = 0; i < n; i++) {
+      var id = {{{ makeGetValue('ids', 'i*4', 'i32') }}};
+      var query = GL.queries[id];
```
`var query = GL.timerQueriesEXT[id];`

gouletr commented:
@juj Ready for review. I updated the branch to account for the comments you said, and also implemented `glQueryCounterEXT`. I also implemented the 4 versions of `glGetQueryObject[i/ui/i64/ui64]vEXT`, please take a look, especially the return value (if that's the proper way).

Cheers!

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1403,180 @@ var LibraryGL = {
+  },
+
+  glGetQueryObjectuivEXT__sig: 'viii',
+  glGetQueryObjectuivEXT: function(id, pname, params) {
```
The function `glGetQueryObjectuivEXT` is identical to `glGetQueryObjectivEXT`, so let's safe here a little and do  `glGetQueryObjectuivEXT: 'glGetQueryObjectivEXT'`. The prints in that function could read e.g. "glGetQueryObject(u)ivEXT" to be generic.

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1403,180 @@ var LibraryGL = {
+  },
+
+  glGetQueryObjectui64vEXT__sig: 'viii',
+  glGetQueryObjectui64vEXT: function(id, pname, params) {
```
This is identical to `glGetQueryObjecti64vEXT`, so let's do `glGetQueryObjectui64vEXT: 'glGetQueryObjecti64vEXT'`. The makeSetValue here is ok, which is special in this case because the WebGL disjoint timer query extension is restricted to only ever return an "64-bit int that in double" (which is only up to a 53-bit integer) and not an arbitrary int64.

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1403,180 @@ var LibraryGL = {
+      var query = GL.timerQueriesEXT[id];
+      if (!query) continue; // GL spec: "unused names in ids are ignored, as is the name zero."
+      context.disjointTimerQueryExt.deleteQueryEXT(query);
+      query.name = 0;
```
I think it's good to just delete the `query.name = 0;` line.

juj commented on src/library_gl.js:
```diff
@@ -1400,6 +1403,180 @@ var LibraryGL = {
+  glGenQueriesEXT__sig: 'vii',
+  glGenQueriesEXT: function(n, ids) {
+    for (var i = 0; i < n; i++) {
+      var query = context.disjointTimerQueryExt.createQueryEXT();
```
To be compatible with the closure compiler minification, this should be `context.disjointTimerQueryExt['createQueryEXT']();` instead. Likewise for all cases below where a member function of `context.disjointTimerQueryExt` is accessed. Leaving this change up to you.

juj commented:
Looks great now! Minor comments, hoping to save a bit of code duplication by merging those two functions, and if you'd like, you can remove the silly `query.name = 0;` lines in the extension and from the original WebGL 2 version of the code.

Also, the `timerQueryExtension['foo']()` notation would be preferred over `timerQueryExtension.foo()` to access the members of the extension object, because closure compiler does not know that `timerQueryExtension` is a browser built-in provided API.

gouletr commented:
Thanks @juj I'm working on applying those change, and I started testing the latest version and I get the following error when using any of those function:
```
exception thrown: ReferenceError: context is not defined,ReferenceError: context is not defined
```
Why is the context member not available in the functions?

juj commented:
Oh, yeah, that's wrong - I was completely blind to that one. The currently active GL context is available under the name `GLctx` in the query-related functions, not `context`. Only in the creation function with `context.disjointTimerQueryExt = GLctx.getExtension("EXT_disjoint_timer_query");` it's correct.

gouletr commented:
Storing the extension into `GLctx.disjointTimerQueryExt` instead fixed the issue. Not sure what's the deal with `context` not being available globally.

gouletr commented:
Oh I see, I'll try that.

juj commented:
Err, yeah, I mean `GLctx.disjointTimerQueryExt = GLctx.getExtension("EXT_disjoint_timer_query");` is correct. `context` is a local variable in that function, which refers to the structure object of GL context plus its canvas. We don't create a global JS variable `context` anywhere, but we do have that global JS variable `GLctx` which gets changed to point to the currently active GL context when any of the `glMakeCurrent` family of functions is called.

gouletr commented:
Ok ready for review again. Cheers! :)

juj commented:
Perfect!

pyalot commented:
Please note that the consensus regarding querying has evolved in the following fashion:

* In WebGL1 there is EXT_disjoint_timer_query which provides the functions (beginQueryEXT etc.) as well as the enums.
* In WebGL2 there is EXT_disjoint_timer_query_webgl2 which provides just the enums and relies on the WebGL2 core functions (beginQuery etc.)

It is unlikely that another conflicting extension will be introduced to WebGL1, so it should be safe to assign the query related functions to the ES 3.0 entry points in a WebGL1 context, and in a WebGL2 context just use the enums and rely on the core functions.

gouletr commented:
@pyalot Thanks. It doesn't seem like this `EXT_disjoint_timer_query_webgl2` is ready or available in browsers right now, so for the moment, and until this is fully implemented, we can get proper GPU timings by using EXT versions of query functions. Until then I suggest we keep it like this, we need those timings. Cheers!

mosra commented:
*(Commenting here so people subscribed to this thread get notified.)*

I just opened https://github.com/emscripten-core/emscripten/pull/9652, adding working `EXT_disjoint_timer_query_webgl2` support.
